new ESM implementation (#26745)
nodejs/node: doc/api/esm.md

> Stability: 1 - Experimental

## Introduction

> ECMAScript modules are the official standard format to package JavaScript code

Reviewer: Link to the spec here?

Author: Fixed

code for reuse. Modules are defined using a variety of [`import`][] and
[`export`][] statements.

Node.js fully supports ECMAScript modules as they are currently specified and
provides limited interoperability between them and the existing module format,
[CommonJS][].

> <!--name=esm-->

Author: fixed in 5833074f33

Node.js contains support for ES Modules based upon the

> [Node.js EP for ES Modules][] and the [ecmascript-modules implementation][].

Author: fixed in 3d5152ecaa

Expect major changes in the implementation including interoperability support,
specifier resolution, and default behavior.

## Enabling

<!-- type=misc -->

The `--experimental-modules` flag can be used to enable support for
ECMAScript modules (ES modules).

## Running Node.js with an ECMAScript Module

There are a few ways to start Node.js with an ES module as its input.

### Initial entry point with an <code>.mjs</code> extension

A file ending with `.mjs` passed to Node.js as an initial entry point will be
loaded as an ES module.

```sh
node --experimental-modules my-app.mjs
```

> ### <code>--type=module</code> / <code>-m</code> flag

Reviewer 2: ```suggestion
### <code>--type=module</code> flag
```

as an ES module input sent in via `--eval` or `--print` (or `-e` or `-p`) or
piped to Node.js via `STDIN`.

```sh
node --experimental-modules --type=module --eval \
  "import { sep } from 'path'; console.log(sep);"

echo "import { sep } from 'path'; console.log(sep);" | \
  node --experimental-modules --type=module
```

### <code>package.json</code> <code>"type"</code> field

Files ending with `.js` or `.mjs`, or lacking any extension,
will be loaded as ES modules when the nearest parent `package.json` file
contains a top-level field `"type"` with a value of `"module"`.

The nearest parent `package.json` is defined as the first `package.json` found
when searching in the current folder, that folder’s parent, and so on up
until the root of the volume is reached.

<!-- eslint-skip -->
```js
// package.json
{
  "type": "module"
}
```

```sh
# In same folder as above package.json
node --experimental-modules my-app.js # Runs as ES module
```

If the nearest parent `package.json` lacks a `"type"` field, or contains
`"type": "commonjs"`, extensionless and `.js` files are treated as CommonJS.
If the volume root is reached and no `package.json` is found,

> Node.js defers to the deafult, a `package.json` with no `"type"`

Reviewer 3: ```suggestion
Node.js defers to the default, a `package.json` with no `"type"`
```

--experimental-modules my-app.js`) but also to files referenced by `import`
statements and `import()` expressions.

```js
// my-app.js, in an ES module package scope because there is a package.json
// file in the same folder with "type": "module"

import './startup/init.js';
// Loaded as ES module since ./startup contains no package.json file,
// and therefore inherits the ES module package scope from one level up

import 'commonjs-package';
// Loaded as CommonJS since ./node_modules/commonjs-package/package.json
// lacks a "type" field or contains "type": "commonjs"

import './node_modules/commonjs-package/index.js';
// Loaded as CommonJS since ./node_modules/commonjs-package/package.json
// lacks a "type" field or contains "type": "commonjs"
```

Files ending with `.mjs` are always loaded as ES modules regardless of package
scope.

Files ending with `.cjs` are always loaded as CommonJS regardless of package
scope.

```js
import './legacy-file.cjs';
// Loaded as CommonJS since .cjs is always loaded as CommonJS

import 'commonjs-package/src/index.mjs';
// Loaded as ES module since .mjs is always loaded as ES module
```

> You can use the `.mjs` and `.cjs` extensions to mix types within the same

Reviewer: Please avoid "you" in the docs

```suggestion
The `.mjs` and `.cjs` extensions may be used to mix types within the same
```

* `url` {string} The absolute `file:` URL of the module.

## Differences Between ES Modules and CommonJS

### Mandatory file extensions

> You must provide a file extension when using the `import` keyword. Directory

Reviewer: ```suggestion
A file extension must be provided when using the `import` keyword. Directory
```

fs.readFileSync === readFileSync;
```

## Experimental JSON Modules

**Note: This API is still being designed and is subject to change.**.

Currently importing JSON modules are only supported in the `commonjs` mode
and are loaded using the CJS loader. [WHATWG JSON modules][] are currently
being standardized, and are experimentally supported by including the
additional flag `--experimental-json-modules` when running Node.js.

When the `--experimental-json-modules` flag is included both the
`commonjs` and `module` mode will use the new experimental JSON
loader. The imported JSON only exposes a `default`, there is no
support for named exports. A cache entry is created in the CommonJS
cache, to avoid duplication. The same object will be returned in
CommonJS if the JSON module has already been imported from the
same path.

Assuming an `index.js` with

<!-- eslint-skip -->
```js
import packageConfig from './package.json';
```

> You will need to include the `--experimental-json-modules` flag for the module

Reviewer: ```suggestion
The `--experimental-json-modules` flag is needed for the module
```

> 1. Otherwise,
>    1. If _isMain_ is **true** or _url_ ends in _".js"_, _".json"_ or
>       _".node"_, then
>       1. Return _"commonjs"_.
>    1. Throw an _Unsupported File Extension_ error.

READ_PACKAGE_SCOPE(_url_)
> 1. Let _scopeURL_ be _url_.
> 1. While _scopeURL_ is not the file system root,
>    1. Let _pjson_ be the result of **READ_PACKAGE_JSON**(_scopeURL_).
>    1. If _pjson_ is not **null**, then
>       1. Return _pjson_.
>    1. Set _scopeURL_ to the parent URL of _scopeURL_.
> 1. Return **null**.

READ_PACKAGE_JSON(_packageURL_)
> 1. Let _pjsonURL_ be the resolution of _"package.json"_ within _packageURL_.
> 1. If the file at _pjsonURL_ does not exist, then
>    1. Return **null**.
> 1. If the file at _packageURL_ does not parse as valid JSON, then
>    1. Throw an _Invalid Package Configuration_ error.
> 1. Return the parsed JSON source of the file at _pjsonURL_.

</details>

### Customizing ESM specifier resolution algorithm

The current specifier resolution does not support all default behavior of
the CommonJS loader. One of the behavior differences is automatic resolution
of file extensions and the ability to import directories that have an index
file.

> By using the `--es-module-specifier-resolution=[mode]` flag you can customize

Reviewer: ```suggestion
The `--es-module-specifier-resolution=[mode]` flag can be used to customize
```

then be called at the exact point of module evaluation order for that module
in the import tree.

## Resolution Algorithm

### Features

The resolver has the following properties:

* FileURL-based resolution as is used by ES modules
* Support for builtin module loading
* Relative and absolute URL resolution
* No default extensions
* No folder mains
* Bare specifier package resolution lookup through node_modules

### Resolver Algorithm

The algorithm to load an ES module specifier is given through the
**ESM_RESOLVE** method below. It returns the resolved URL for a
module specifier relative to a parentURL, in addition to the unique module
format for that resolved URL given by the **ESM_FORMAT** routine.

The _"module"_ format is returned for an ECMAScript Module, while the
_"commonjs"_ format is used to indicate loading through the legacy
CommonJS loader. Additional formats such as _"wasm"_ or _"addon"_ can be
extended in future updates.

> In the following algorithms, all subroutine errors are propogated as errors

Reviewer 4: ```suggestion
In the following algorithms, all subroutine errors are propagated as errors
```

> 1. Let _format_ be the result of **ESM_FORMAT**(_resolvedURL_, _isMain_).
> 1. Load _resolvedURL_ as module format, _format_.

PACKAGE_RESOLVE(_packageSpecifier_, _parentURL_)
> 1. Let _packageName_ be *undefined*.
> 1. Let _packageSubpath_ be *undefined*.
> 1. If _packageSpecifier_ is an empty string, then
>    1. Throw an _Invalid Specifier_ error.
> 1. If _packageSpecifier_ does not start with _"@"_, then
>    1. Set _packageName_ to the substring of _packageSpecifier_ until the
>       first _"/"_ separator or the end of the string.
> 1. Otherwise,
>    1. If _packageSpecifier_ does not contain a _"/"_ separator, then
>       1. Throw an _Invalid Specifier_ error.
>    1. Set _packageName_ to the substring of _packageSpecifier_
>       until the second _"/"_ separator or the end of the string.
> 1. Let _packageSubpath_ be the substring of _packageSpecifier_ from the
>    position at the length of _packageName_ plus one, if any.
> 1. Assert: _packageName_ is a valid package name or scoped package name.
> 1. Assert: _packageSubpath_ is either empty, or a path without a leading
>    separator.
> 1. If _packageSubpath_ contains any _"."_ or _".."_ segments or percent

> >    encoded strings for _"/"_ or _"\"_ then,

Reviewer 4: ```suggestion
>    encoded strings for _"/"_ or _"\\"_ then,
```

Reviewer 5: Thanks.

>    1. Throw an _Invalid Specifier_ error.
> 1. If _packageSubpath_ is empty and _packageName_ is a Node.js builtin
>    module, then
>    1. Return the string _"node:"_ concatenated with _packageSpecifier_.
> 1. While _parentURL_ is not the file system root,

> >    1. Set _parentURL_ to the parent folder URL of _parentURL_.

Reviewer 4: Sorry if I am confused, but does this mean we do not start searching in the `node_modules` subfolder of the current folder, but in the `node_modules` subfolder of the parent folder right away?

Reviewer 5: Because `parentURL` is a file path, the parent folder URL does the direct parent `node_modules` change.

We could possibly clarify this by calling it `parentDirURL`.

Reviewer 5: I actually just refactored the wording of the resolution to get this one right.

The current specifier resolution does not support all default behavior of
the CommonJS loader. One of the behavior differences is automatic resolution
of file extensions and the ability to import directories that have an index
file.

The `--es-module-specifier-resolution=[mode]` flag can be used to customize
the extension resolution algorithm. The default mode is `explicit`, which
requires the full path to a module be provided to the loader. To enable the
automatic extension resolution and importing from directories that include an
index file use the `node` mode.

```bash
$ node --experimental-modules index.mjs
success!
$ node --experimental-modules index #Failure!
Error: Cannot find module
$ node --experimental-modules --es-module-specifier-resolution=node index
success!
```

[`import`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import
[`export`]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export
[CommonJS]: modules.html
[Node.js EP for ES Modules]: https://github.com/nodejs/node-eps/blob/master/002-es-modules.md
[dynamic instantiate hook]: #esm_dynamic_instantiate_hook

> [`module.createRequireFromPath()`]: modules.html#modules_module_createrequirefrompath_filename

Author: fixed in 26b208a335
